File: archconvnets/unsupervised/bp.c
// compile with: gcc bp.c -o bp -std=c99 -O3 -lm
// derivations in archconvnets/unsupervised/docs/darren_scan.pdf
// all matrices stored linearly in memory (column-major order);
// the [X]_IND(A, B, C, D) macros convert subscript indices into the linear index

//#define DEBUG 1 // undefining this removes several index bounds checks, speeding up code execution

#include "bp_defs.h"
#include "bp_conv.c" // conv()
#include "bp_max_pool.c" // max_pool()

float * imgs;
float * pred; // label predictions [number categories x N_IMGS]
int output_sz1, output_sz2, output_sz3; // size in px of each conv. output spatial size
int max_output_sz1, max_output_sz2, max_output_sz3; // size in px of each pooling output spatial size
float *F1, *F2, *F3, *FL; // filters
float *output1, *output2, *output3; // conv outputs
float *max_output1, *max_output2, *max_output3; // max pooling outputs
int *switch_output1, *switch_output2, *switch_output3; // max pooling switches (indices into previous conv layer)
float * Y;

#include "bp_return_px.c" // return_px()
#include "bp_compute_preds_cost.c" // compute_preds(), compute_cost()

int main(){
	srand(time(NULL));
	int rand_ind;
	unsigned t_start;
	float grad;
		
	output_sz1 = floor((IMG_SZ - s1) / STRIDE1);
	max_output_sz1 = floor((output_sz1 - POOL_SZ) / POOL_STRIDE);

	output_sz2 = floor((max_output_sz1 - s2));
	max_output_sz2 = floor((output_sz2 - POOL_SZ) / POOL_STRIDE);
	
	output_sz3 = floor((max_output_sz2 - s3));
	max_output_sz3 = floor((output_sz3 - POOL_SZ) / POOL_STRIDE);

	MALLOC(output1, n1 * output_sz1 * output_sz1 * N_IMGS * 2, sizeof(float));
	MALLOC(max_output1, n1 * max_output_sz1 * max_output_sz1 * N_IMGS *2, sizeof(float));
	MALLOC(switch_output1, n1 * max_output_sz1 * max_output_sz1 * N_IMGS*2, sizeof(float));

	MALLOC(output2, n2 * output_sz2 * output_sz2 * N_IMGS*2, sizeof(float));
	MALLOC(max_output2, n2 * max_output_sz2 * max_output_sz2 * N_IMGS*2, sizeof(float));
        MALLOC(switch_output2, n2 * max_output_sz2 * max_output_sz2 * N_IMGS*2, sizeof(float));

	MALLOC(output3, n3 * output_sz3 * output_sz3 * N_IMGS*2, sizeof(float))
	MALLOC(max_output3, n3 * max_output_sz3 * max_output_sz3 * N_IMGS*2, sizeof(float));
        MALLOC(switch_output3, n3 * max_output_sz3 * max_output_sz3 * N_IMGS*2, sizeof(float));

	MALLOC(pred, N_C * N_IMGS*2, sizeof(float));

	MALLOC_RAND(F1, n1 * 3 * s1_2*2, sizeof(float));
	MALLOC_RAND(F2, n2 * n1 * s2_2*2, sizeof(float));
	MALLOC_RAND(F3, n3 * n2 * s3_2*2, sizeof(float));
	MALLOC_RAND(FL, N_C * n3 * max_output_sz3 * max_output_sz3*2, sizeof(float));
	MALLOC_RAND(imgs, 3 * IMG_SZ2 * N_IMGS*2, sizeof(float));
	MALLOC_RAND(Y, N_C * N_IMGS*2, sizeof(float));
	
	printf("szs: %i %i, %i %i, %i %i\n", output_sz1, max_output_sz1, output_sz2, max_output_sz2, output_sz3, max_output_sz3);
	
	//////////////////////////////////////
	// compute model outputs with current filters
	for(int step=0; step < 100; step++){ //gradient steps
	t_start = (unsigned)time(NULL);
	conv(F1, output1, n1, 3, s1, IMG_SZ, imgs, STRIDE1, output_sz1);
	max_pool(output1, max_output1, switch_output1, n1, output_sz1, max_output_sz1);

	conv(F2, output2, n2, n1, s2, output_sz1, max_output1, 1, output_sz2);
	max_pool(output2, max_output2, switch_output2, n2, output_sz2, max_output_sz2);

	conv(F3, output3, n3, n2, s3, output_sz2, max_output2, 1, output_sz3);
	max_pool(output3, max_output3, switch_output3, n3, output_sz3, max_output_sz3);

	compute_preds();

	printf("forward pass took: %i sec\n", (unsigned)time(NULL) - t_start);
	
	//////////////////////
	// index loop variables for gradient computations
	// variables ending in "_" are indices the gradient is taken wrt
	int f1, f2, f3, f1_ = 1, f2_ = 1, f3_ = 1;
	int cat, cat_=3;
	int a1_x, a1_y, a1_x_ = 4, a1_y_ = 6;
	int channel, channel_ = 0;
	int temp_ind;
	float temp_F_prod_all, temp_F31_prod, temp_F32_prod, temp_F321_prod, temp_F32_prod_px, temp_F3, temp_F2, temp_F21_prod;
	int z1, z2, z1_ = 0, z2_ = 1;
	int a3_x,a3_y, a3_x_,a3_y_, a2_x, a2_y, a2_x_, a2_y_, img;
	
	
	///////////////////////////////////////////////
	// deriv F1: wrt f1_, a1_x_, a1_y_, channel_
	
	t_start = (unsigned)time(NULL);
	for(f1_ = 0; f1_ < 3; f1_++){ //todo: loop over a1_x_, a1_y_, channel_
		grad = 0;
		for(f3=0; f3 < n3; f3++){
		for(f2=0; f2 < n2; f2++){
		for(a3_x=0; a3_x < s3; a3_x++){ for(a3_y=0; a3_y < s3; a3_y++){
			F3_IND_DBG(f3, f2, a3_x, a3_y)
		
			temp_F3 = F3[F3_IND(f3, f2, a3_x, a3_y)];
			
	    	for(a2_x=0; a2_x < s2; a2_x++){ for(a2_y=0; a2_y < s2; a2_y++){
			
				F2_IND_DBG(f2, f1_, a2_x, a2_y)
			
				temp_F32_prod = temp_F3 * F2[F2_IND(f2, f1_, a2_x, a2_y)];
				
				for(img=0; img < N_IMGS; img++){
				for(z1=0; z1 < max_output_sz3; z1++){ for(z2=0; z2 < max_output_sz3; z2++){ 
				
					temp_F32_prod_px = temp_F32_prod * 
						return_px(f1_, f2, f3, z1, z2, a3_x, a3_y, a2_x, a2_y, a1_x_, a1_y_, channel_, img); // return_px(): "X" in the derivations
					for(cat = 0; cat < N_C; cat++){
						FL_IND_DBG(cat, f3, z1, z2)
				
						temp_F_prod_all = FL[FL_IND(cat, f3, z1, z2)] * temp_F32_prod_px;
					
						Y_IND_DBG(cat,img);
						temp_ind = Y_IND(cat,img);
						// supervised term:
						//... sigma approximations ...
						grad -= temp_F_prod_all * Y[temp_ind];
					
						// unsupervised term:
						grad +=  temp_F_prod_all * pred[temp_ind];
		}}}}}}}}}}
		F1_IND_DBG(f1_, channel_, a1_x_, a1_y_)
			
		F1[F1_IND(f1_, channel_, a1_x_, a1_y_)] -= eps_F1*2*grad;
	}
	printf("F1 grad: %i sec\n", (unsigned)time(NULL) - t_start);
	
	///////////////////////////////////////////////
	// deriv FL: wrt cat_, f3_, z1_, z2_

	t_start = (unsigned)time(NULL);
	for(f3_=0; f3_ < 2*6; f3_++){ // todo: for loop over cat_, z1_, z2_
		grad = 0;
		for(f1=0; f1 < n1; f1++){  
		for(f2=0; f2 < n2; f2++){
		for(a3_x=0; a3_x < s3; a3_x++){ for(a3_y=0; a3_y < s3; a3_y++){
		   F3_IND_DBG(f3_,f2, a3_x, a3_y)
		   
		   temp_F3 = F3[F3_IND(f3_,f2, a3_x, a3_y)];
		   
			for(a2_x=0; a2_x < s2; a2_x++){ for(a2_y=0; a2_y < s2; a2_y++){
			
				F2_IND_DBG(f2, f1, a2_x, a2_y)
				
				temp_F32_prod = temp_F3 * F2[F2_IND(f2, f1_, a2_x, a2_y)];
				
				for(a1_x=0; a1_x < s1; a1_x++){ for(a1_y=0; a1_y < s1; a1_y++){
				
					for(channel=0; channel < 3; channel++){
						F1_IND_DBG(f1, channel, a1_x, a1_y)
						
						temp_F321_prod = temp_F32_prod * F1[F1_IND(f1, channel, a1_x, a1_y)];
						
						for(img=0; img < N_IMGS; img++){
								temp_F_prod_all = temp_F321_prod *
									return_px(f1, f2, f3_, z1_, z2_, a3_x, a3_y, a2_x, a2_y, a1_x, a1_y, channel, img); // return_px(): "X" in the derivations
								
								Y_IND_DBG(cat_,img);
								
								temp_ind = Y_IND(cat_,img);
								// supervised term:
								//... sigma approximations ...
								grad -= temp_F_prod_all * Y[temp_ind];

								// unsupervised term:
								grad +=  temp_F_prod_all * pred[temp_ind];
						}
						
		}}}}}}}}}
		FL_IND_DBG(cat_, f3_, z1_, z2_)
		FL[FL_IND(cat_, f3_, z1_, z2_)] -= eps_FL*2*grad;
	}
	printf("F1 grad: %i sec\n", (unsigned)time(NULL) - t_start);
	
	
	///////////////////////////////////////////////
	// deriv F2: wrt f2_, f1_, a2_x_, a2_y_
	
	t_start = (unsigned)time(NULL);
	for(f2_=0; f2_ < 8; f2_++){
		grad = 0;
		for(f3=0; f3 < n3; f3++){
		for(a3_x=0; a3_x < s3; a3_x++){ for(a3_y=0; a3_y < s3; a3_y++){
		   F3_IND_DBG(f3, f2_, a3_x, a3_y)
		   
		   temp_F3 = F3[F3_IND(f3,f2_, a3_x, a3_y)];
		   
		   for(a1_x=0; a1_x < s1; a1_x++){ for(a1_y=0; a1_y < s1; a1_y++){
				
			for(channel=0; channel < 3; channel++){
				F1_IND_DBG(f1_, channel, a1_x, a1_y)
					
				temp_F31_prod = temp_F3 * F1[F1_IND(f1_, channel, a1_x, a1_y)];
						
				for(img=0; img < N_IMGS; img++){
					for(z1=0; z1 < max_output_sz3; z1++){ for(z2=0; z2 < max_output_sz3; z2++){ 
				
						temp_F_prod_all = temp_F31_prod *
									return_px(f1_, f2_, f3, z1, z2, a3_x, a3_y, a2_x_, a2_y_, a1_x, a1_y, channel, img); // return_px(): "X" in the derivations
						for(cat=0; cat < N_C; cat ++){	
							Y_IND_DBG(cat,img);
						
							temp_ind = Y_IND(cat,img);
							// supervised term:
							//... sigma approximations ...
							grad -= temp_F_prod_all * Y[temp_ind];
		
							// unsupervised term:
							grad +=  temp_F_prod_all * pred[temp_ind];
						}
					}
						
		}}}}}}}}
		F2_IND_DBG(f2_, f1_, a2_x_, a2_y_)
		F2[F2_IND(f2_, f1_, a2_x_, a2_y_)] -= eps_F2*2*grad;
	}
	printf("F2 grad: %i sec\n", (unsigned)time(NULL) - t_start);
	
	
	///////////////////////////////////////////////
	// deriv F3: wrt f3_, f2_, a3_x_, a3_y_
	//?
	t_start = (unsigned)time(NULL);
	for(f3_=0; f3_ < n3; f3_++){ for(f2_=0; f2_ < 3; f2_++){
		grad = 0;
		for(a2_x=0; a2_x < s2; a2_x++){ for(a2_y=0; a2_y < s2; a2_y++){
		   F2_IND_DBG(f2_, f1, a2_x, a2_y)
		   
		   temp_F2 = F2[F2_IND(f2_, f1, a2_x, a2_y)];
		   
		   for(a1_x=0; a1_x < s1; a1_x++){ for(a1_y=0; a1_y < s1; a1_y++){
				
			for(channel=0; channel < 3; channel++){
				F1_IND_DBG(f1, channel, a1_x, a1_y)
					
				temp_F21_prod = temp_F2 * F1[F1_IND(f1, channel, a1_x, a1_y)];
						
				for(img=0; img < N_IMGS; img++){
					for(z1=0; z1 < max_output_sz3; z1++){ for(z2=0; z2 < max_output_sz3; z2++){ 
				
						temp_F_prod_all = temp_F21_prod *
									return_px(f1, f2_, f3_, z1, z2, a3_x_, a3_y_, a2_x, a2_y, a1_x, a1_y, channel, img); // return_px(): "X" in the derivations
						for(cat=0; cat < N_C; cat ++){	
							Y_IND_DBG(cat,img);
						
							temp_ind = Y_IND(cat,img);
							// supervised term:
							//... sigma approximations ...
							grad -= temp_F_prod_all * Y[temp_ind];
		
							// unsupervised term:
							grad +=  temp_F_prod_all * pred[temp_ind];
						}
					}
						
		}}}}}}}
		F3_IND_DBG(f3_, f2_, a3_x_, a3_y_)
		F3[F3_IND(f3_, f2_, a3_x_, a3_y_)] -= eps_F3*2*grad;
	}}
	printf("F3 grad: %i sec\n", (unsigned)time(NULL) - t_start);
	
	printf("%f, cost %f, F1 grad: %i sec\n", grad, compute_cost(), (unsigned)time(NULL) - t_start);
	} // gradient steps
	return 0;
}

File: archconvnets/unsupervised/bp.py
#import numpy as npd
import time
import numpy as np
from archconvnets.unsupervised.conv import conv_block
from archconvnets.unsupervised.pool_inds import max_pool_locs
from archconvnets.unsupervised.compute_L1_grad import L1_grad
from archconvnets.unsupervised.compute_L2_grad import L2_grad
from archconvnets.unsupervised.compute_L3_grad import L3_grad
from archconvnets.unsupervised.compute_FL_grad import FL_grad
from scipy.io import savemat

F1_scale = 0.01 # std of init normal distribution
F2_scale = 0.01
F3_scale = 0.01
FL_scale = 0.01

eps_F1 = 1e-2#1e-7
eps_F2 = 1e-2#1e-7
eps_F3 = 1e-3#1e-8
eps_FL = 1e-2

POOL_SZ = 3
POOL_STRIDE = 2
STRIDE1 = 1 # layer 1 stride
N_IMGS = 2 # batch size
N_TEST_IMGS = 100
IMG_SZ = 42 # input image size (px)

N = 4 
n1 = N # L1 filters
n2 = N # ...
n3 = N

s3 = 3 # L1 filter size (px)
s2 = 5 # ...
s1 = 5

N_C = 10 # number of categories

output_sz1 = len(range(0, IMG_SZ - s1 + 1, STRIDE1))
max_output_sz1  = len(range(0, output_sz1-POOL_SZ, POOL_STRIDE))

output_sz2 = max_output_sz1 - s2 + 1
max_output_sz2  = len(range(0, output_sz2-POOL_SZ, POOL_STRIDE))

output_sz3 = max_output_sz2 - s3 + 1
max_output_sz3  = len(range(0, output_sz3-POOL_SZ, POOL_STRIDE))

F1 = np.random.normal(scale=F1_scale, size=(n1, 3, s1, s1))
F2 = np.random.normal(scale=F2_scale, size=(n2, n1, s2, s2))
F3 = np.random.normal(scale=F3_scale, size=(n3, n2, s3, s3))
FL = np.random.normal(scale=FL_scale, size=(N_C, n3, max_output_sz3, max_output_sz3))

imgs_mean = np.load('/home/darren/cifar-10-py-colmajor/batches.meta')['data_mean']

z = np.load('/home/darren/cifar-10-py-colmajor/data_batch_1')

err = []
class_err = []
err_test = []
class_err_test = []
for step in range(np.int(1e7)):
	########################## compute test err
	# load imgs
	x = z['data'] - imgs_mean
	x = x.reshape((3, 32, 32, 10000))
	x = x[:,:,:,10000-N_TEST_IMGS:]

	l = np.zeros((N_TEST_IMGS, N_C),dtype='int')
	l[np.arange(N_TEST_IMGS),np.asarray(z['labels'])[10000-N_TEST_IMGS:].astype(int)] = 1
	Y = np.double(l.T)

	imgs_pad = np.zeros((3, IMG_SZ, IMG_SZ, N_TEST_IMGS))
	imgs_pad[:,5:5+32,5:5+32] = x

	# forward pass
	t_test_forward_start = time.time()
	conv_output1 = conv_block(F1.transpose((1,2,3,0)), imgs_pad, stride=STRIDE1)
	max_output1, output_switches1_x, output_switches1_y = max_pool_locs(conv_output1)

	conv_output2 = conv_block(F2.transpose((1,2,3,0)), max_output1)
	max_output2, output_switches2_x, output_switches2_y = max_pool_locs(conv_output2)

	conv_output3 = conv_block(F3.transpose((1,2,3,0)), max_output2)
	max_output3, output_switches3_x, output_switches3_y = max_pool_locs(conv_output3)

	pred = np.dot(FL.reshape((N_C, n3*max_output_sz3**2)), max_output3.reshape((n3*max_output_sz3**2, N_TEST_IMGS)))
	err_test.append(np.sum((pred - Y)**2))
	class_err_test.append(1-np.float(np.sum(np.argmax(pred,axis=0) == np.argmax(Y,axis=0)))/N_TEST_IMGS)
	
	t_test_forward_start = time.time() - t_test_forward_start

	################### compute train err
	# load imgs
	x = z['data'] - imgs_mean
	x = x.reshape((3, 32, 32, 10000))
	x = x[:,:,:,step*N_IMGS:(step+1)*N_IMGS]

	l = np.zeros((N_IMGS, N_C),dtype='int')
	l[np.arange(N_IMGS),np.asarray(z['labels'])[step*N_IMGS:(step+1)*N_IMGS].astype(int)] = 1
	Y = np.double(l.T)

	imgs_pad = np.zeros((3, IMG_SZ, IMG_SZ, N_IMGS))
	imgs_pad[:,5:5+32,5:5+32] = x


	# forward pass
	t_forward_start = time.time()
	conv_output1 = conv_block(F1.transpose((1,2,3,0)), imgs_pad, stride=STRIDE1)
	max_output1, output_switches1_x, output_switches1_y = max_pool_locs(conv_output1)

	conv_output2 = conv_block(F2.transpose((1,2,3,0)), max_output1)
	max_output2, output_switches2_x, output_switches2_y = max_pool_locs(conv_output2)

	conv_output3 = conv_block(F3.transpose((1,2,3,0)), max_output2)
	max_output3, output_switches3_x, output_switches3_y = max_pool_locs(conv_output3)

	pred = np.dot(FL.reshape((N_C, n3*max_output_sz3**2)), max_output3.reshape((n3*max_output_sz3**2, N_IMGS)))
	err.append(np.sum((pred - Y)**2))
	class_err.append(1-np.float(np.sum(np.argmax(pred,axis=0) == np.argmax(Y,axis=0)))/N_IMGS)

	output_switches1_x *= STRIDE1
	output_switches1_y *= STRIDE1
	
	t_forward_start = time.time() - t_forward_start
	
	########### F1 deriv wrt f1_, a1_x_, a1_y_, channel_

	t_grad_start = time.time()
	grad = L1_grad(F1, F2, F3, FL, output_switches3_x, output_switches3_y, output_switches2_x, output_switches2_y, output_switches1_x, output_switches1_y, s1, s2, s3, pred, Y, imgs_pad)
	F1 -= eps_F1 * grad
	
	########### F2 deriv wrt f2_, f1_, a2_x_, a2_y_

	grad = L2_grad(F1, F2, F3, FL, output_switches3_x, output_switches3_y, output_switches2_x, output_switches2_y, output_switches1_x, output_switches1_y, s1, s2, s3, pred, Y, imgs_pad)
	F2 -= eps_F2 * grad
	
	########### F3 deriv wrt f3_, f2_, a3_x_, a3_y_

	grad = L3_grad(F1, F2, F3, FL, output_switches3_x, output_switches3_y, output_switches2_x, output_switches2_y, output_switches1_x, output_switches1_y, s1, s2, s3, pred, Y, imgs_pad)
	F3 -= eps_F3 * grad
	
	########### FL deriv wrt cat_, f3_, z1_, z2_

	t_grad_start = time.time()
	grad = FL_grad(F1, F2, F3, FL, output_switches3_x, output_switches3_y, output_switches2_x, output_switches2_y, output_switches1_x, output_switches1_y, s1, s2, s3, pred, Y, imgs_pad)
	FL -= eps_FL * grad
	
	#######################################
	
	savemat('/home/darren/cifar_F1t.mat', {'F1': F1, 'F2': F2, 'F3':F3, 'eps_F3': eps_F3, 'eps_F2': eps_F2, 'step': step, 'eps_F1': eps_F1, 'N_IMGS': N_IMGS, 'N_TEST_IMGS': N_TEST_IMGS,'err_test':err_test,'err':err,'class_err':class_err,'class_err_test':class_err_test})
	print step, err_test[-1], class_err_test[-1], err[-1], class_err[-1], time.time() - t_grad_start

